### ISSUE TITLE
PLATO: Pre-trained Dialogue Generation Model with Discrete Latent Variable

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#3]
# 논문 정보 (URL, 저자 등)

[https://arxiv.org/abs/1910.07931](https://arxiv.org/abs/1910.07931)

Baidu

# 개요

- 대화 생성을 위한 사전학습 모델입니다.
- BERT를 대화 생성 태스크에 직접적으로 적용하는 데에는 몇가지 문제점이 있습니다.
    1. 사람 간의 대화에서의 언어적인 특징은 기계 번역, 문서 요약 등의 문장 생성 태스크와는 데이터 분포, 지식 면에서 확연히 다릅니다.
    2. Bidirectional하게 학습된 BERT를 unidirectional한 특성을 지닌 대화 생성 분야에 직접 적용하기가 어렵습니다.
    3. 여타 생성 태스크들과 달리 대화 태스크는 one-to-many한 특성을 갖고 있습니다.
- 본 논문은 위와 같은 문제를 해결하기 위해 다음과 같은 방법론을 제안합니다.
    1. 데이터 분포 차이를 줄이기 위해, Reddit이나 Twitter 등의 대화 데이터를 이용합니다.
    2. Uni-LM의 영향을 받아, uni-, bi-directional processing을 통합하는 학습 방법을 이용합니다.
    3. one-to-many 관계를 모델링하기 위해 discrete latent variable을 이용합니다.

# 모델

- 모델은 총 3가지의 구성 요소로 이루어져 있습니다. 각 구성 요소들간의 관계는 아래의 그림과 같습니다.
    - C: context. 단, knowledge-grounded 일 경우는 context의 뒤에 knowledge를 concat하여 사용합니다.
    - R: response.
    - Z: latent variable. K개의 카테고리컬 잠재 변수로, 각 잠재 변수는 response의 speech act에 대응합니다.

<img width="452" alt="스크린샷 2020-08-29 오후 9 57 35" src="https://user-images.githubusercontent.com/49383323/91637326-a8442200-ea42-11ea-9dcf-33fa905b2da4.png">

- 즉, 구하고자 하는 확률은 P(z|c,r), P(r|c,z) 두 가지로, **response generation** 및 **speech act recognition**의 두가지 태스크를 포함하고 있습니다.
    - 두 가지 태스크를 joint하게 학습하는 통합된 구조는 아래 그림과 같습니다.
    - 기본적으로 Uni-LM의 트랜스포머 구조에 영향을 받은 backbone 구조이고, uni-, bi-directional한 인코딩이 둘 다 가능한 구조입니다.
<img width="925" alt="스크린샷 2020-08-29 오후 9 56 41" src="https://user-images.githubusercontent.com/49383323/91637310-8fd40780-ea42-11ea-8427-36478f54e64d.png">

- Input Representation은 아래와 같습니다.
<img width="898" alt="스크린샷 2020-08-29 오후 9 57 16" src="https://user-images.githubusercontent.com/49383323/91637319-9cf0f680-ea42-11ea-90ca-b21e29d26b30.png">

    - Token Embeddings, Role Embeddings, Turn Embeddings, Position Embeddings으로 구성되어 있습니다.
    - Input 토큰의 맨 처음에는 latent variable z를 입력하고, 각 턴의 끝에는 [EOS]를, response의 시작에는 [BOS]를 붙입니다.
    - Role Embeddings는 기본 A,B의 두턴으로 구성되어 있고, external knowledge를 이용하는 경우 knowledge에는 C를 부여합니다.
    - Turn Embeddings는 Response가 항상 0번이고, Context가 -1,-2,-3,...등을 가져가게 해서 모델이 response를 명시적으로 알수 있게 한 점이 특징입니다.

## 사전학습 (엄밀하게는 Post-training에 가깝습니다.)

- 본 모델을 총 3가지 로스로 학습합니다.
    1. NLL Loss
<img width="433" alt="스크린샷 2020-08-29 오후 9 58 39" src="https://user-images.githubusercontent.com/49383323/91637361-d9245700-ea42-11ea-915b-8315d1bc4af8.png">

    - z는 (c,r)에 대한 latent speech act (dialog act)이고, p(z|c,r)로부터 샘플링합니다.
    - p(z|c,r)는 아래와 같은 식으로부터 구하는데, Figure 2와 같이 첫 토큰 (special mask token)에 FNN을 붙여 예측합니다.

<img width="429" alt="스크린샷 2020-08-29 오후 9 58 44" src="https://user-images.githubusercontent.com/49383323/91637366-dd507480-ea42-11ea-9335-678ef936169b.png">

    1. BOW Loss
        - BOW 로스는 NLL 로스와 다르게, 단어 간의 순서를 무시하여 latent variable이 target response의 global한 정보를 학습할 수 있도록 합니다.
<img width="419" alt="스크린샷 2020-08-29 오후 9 58 48" src="https://user-images.githubusercontent.com/49383323/91637373-e8a3a000-ea42-11ea-891c-07ec0802a632.png">

        - 여기서 f는 아래와 같이 latent variable의 final hidden state를 이용하여 target response 안에 포함된 단어를 non-autoregressive하게 예측하는 함수입니다.

<img width="372" alt="스크린샷 2020-08-29 오후 9 59 38" src="https://user-images.githubusercontent.com/49383323/91637382-f35e3500-ea42-11ea-8ce0-94e88d4cf2af.png">

    2. Response Selection Loss
        - ground truth는 1로 예측하고, random하게 sampling한 negative sample은 0으로 예측하도록 학습합니다.
<img width="445" alt="스크린샷 2020-08-29 오후 9 59 56" src="https://user-images.githubusercontent.com/49383323/91637392-096bf580-ea43-11ea-92fa-5de3b95bf99f.png">

<img width="407" alt="스크린샷 2020-08-29 오후 10 00 15" src="https://user-images.githubusercontent.com/49383323/91637395-0bce4f80-ea43-11ea-9bcc-33de55f3412e.png">

    - 사전 학습은 BERT-BASE 모델로 초기화된 네트워크 파라미터를 베이스로 하여 학습이 진행됩니다.
    - Twitter, Reddit에서 샘플링한 약 8.3M의 샘플을 이용했습니다.
    - speech act recognition, response generation 태스크를 동시에 수행하기 위해서는 네트워크를 2번 통과해야 하므로, 다음과 같이 사전학습을 진행했습니다.
        1. Latent Act Recognition과 Response Selection을 진행
            - p(z|c,r)을 구하고, RS Loss를 구함
        2. Response Generation을 진행
            - NLL Loss와 BOW Loss를 구함
        3. Optimization
            - 전체 네트워크를 업데이트

**하이퍼 파라미터**
- context length:256, response length:50
- transformer layer:12, embedding dimension:768
- batch size:64
- latent variable number K:20
- Adam with lr 5e-5
- V100 8개로 3.5M steps, 총 2주 정도 걸림

# 데이터

- 성능 평가를 위해 총 3개의 오픈 데이터를 이용했습니다.
    1. **Persona-Chat**: 페르소나 프로필을 바탕으로 두 사람이 대화
    2. **Daily Dialog**: 일상 생활에 대한 칫챗
    3. **DSTC7-AVSD**
        - Audio Visual Scene-aware Dialog. 대화 qa 데이터셋. 배경 지식과 문맥을 토대로 대화
        - 텍스트 정보만 이용하는 single-modal 옵션과 audio, visual 정보도 활용하는 multi-modal 옵션이 있음

# 실험

- Evaluation Metric
    - BLEU: 생성 문장의 n-gram이 ground truth 답변에 얼마나 포함이 되어있는지 (precision)
    - Distinct-1/2 : distinct uni- or bi-gram의 수를 전체 생성 단어로 나눈 값 (다양성 판단)
    - Knowledge R/P/F1: 배경 지식에 대한 informativeness

<img width="363" alt="스크린샷 2020-08-29 오후 10 00 48" src="https://user-images.githubusercontent.com/49383323/91637403-1be62f00-ea43-11ea-91c3-ff5466a6ee88.png">

- Human Evaluation
    - 크라우드 소싱으로 fluency, coherence, informativeness, overall에 대해 [0,1,2] 값을 레이블링 받았습니다.
    - 테스트용 세션은 총 100개를 사용했습니다.

**결과**

<img width="895" alt="스크린샷 2020-08-29 오후 10 01 17" src="https://user-images.githubusercontent.com/49383323/91637421-3e784800-ea43-11ea-9cc5-0fb6ff4b81f4.png">
<img width="912" alt="스크린샷 2020-08-29 오후 10 01 26" src="https://user-images.githubusercontent.com/49383323/91637424-40daa200-ea43-11ea-97fb-11e612724d33.png">

- 결과 중 재미있던 건 Reddit 학습 셋에 URL이 포함되어 있어서 그랬는지, 답변으로 그럴싸한 위키피디아 문서 링크를 생성했던 것.

<img width="901" alt="스크린샷 2020-08-29 오후 10 02 10" src="https://user-images.githubusercontent.com/49383323/91637428-4cc66400-ea43-11ea-88a6-a12bd1018dc4.png">

- 1.2와 1.3을 비교해볼 때 bi-directional context information을 이용하는 것이 더 성능이 좋았습니다.
- 그룹 1와 그룹2를 비교해볼 때, Twitter와 Reddit으로 추가 학습을 하는 것이 더 성능에 좋았습니다.
- 2.2와 3.1을 비교해보았을 때, discrete latent variable이 성능을 더 강화하는 효과가 있습니다.

# 논의할만한 내용

- Discrete Latent Variable을 이용하는 방식이 유효한 건 맞아보이지만, 계산량이 두 배로 들기 때문에 꼭 필요한 방법론 같진 않음.

# 관련된 논문
- PLATO-2
